token-client-core/src/main/java/no/nav/security/token/support/client/core/oauth2/OAuth2AccessTokenService.java

```diff
@@ -1,74 +1,90 @@
+    public OAuth2AccessTokenService(OnBehalfOfAssertionResolver assertionResolver,
                                     OnBehalfOfTokenClient onBehalfOfTokenClient,
                                     ClientCredentialsTokenClient clientCredentialsTokenClient) {
-        this.contextHolder = contextHolder;
+        this.assertionResolver = assertionResolver;
         this.onBehalfOfTokenClient = onBehalfOfTokenClient;
         this.clientCredentialsTokenClient = clientCredentialsTokenClient;
     }
 
-    public OAuth2AccessTokenResponse getAccessToken(ClientConfigurationProperties.ClientProperties clientProperties) {
+    public OAuth2AccessTokenResponse getAccessToken(ClientProperties clientProperties) {
         if (clientProperties == null) {
-            throw new OAuth2ClientException("ClientConfigurationProperties.ClientProperties cannot be null");
+            throw new OAuth2ClientException("ClientProperties cannot be null");
         }
-        log.debug("getting access_token with scopes={} for grant={}", clientProperties.getScope(), clientProperties.getGrantType());
+        log.debug("getting access_token with scopes={} for grant={}", clientProperties.getScope(),
+            clientProperties.getGrantType());
         if (isGrantType(clientProperties, OAuth2GrantType.JWT_BEARER)) {
             return getAccessTokenOnBehalfOfAuthenticatedJwtToken(clientProperties);
         } else if (isGrantType(clientProperties, OAuth2GrantType.CLIENT_CREDENTIALS)) {
             return getAccessTokenClientCredentials(clientProperties);
         } else {
-            throw new OAuth2ClientException(String.format("invalid grant-type=%s from OAuth2ClientConfig.OAuth2Client. grant-type not in supported grant-types (%s)",
+            throw new OAuth2ClientException(String.format("invalid grant-type=%s from OAuth2ClientConfig.OAuth2Client" +
+                    ". grant-type not in supported grant-types (%s)",
                 clientProperties.getGrantType().getValue(), SUPPORTED_GRANT_TYPES));
         }
     }
 
-    private OAuth2AccessTokenResponse getAccessTokenOnBehalfOfAuthenticatedJwtToken(ClientConfigurationProperties.ClientProperties clientProperties) {
+    @SuppressWarnings("unused")
+    public Cache<ClientCredentialsGrantRequest, OAuth2AccessTokenResponse> getClientCredentialsGrantCache() {
+        return clientCredentialsGrantCache;
+    }
+
+    public void setClientCredentialsGrantCache(Cache<ClientCredentialsGrantRequest, OAuth2AccessTokenResponse> clientCredentialsGrantCache) {
+        this.clientCredentialsGrantCache = clientCredentialsGrantCache;
+    }
+
+    @SuppressWarnings("unused")
+    public Cache<OnBehalfOfGrantRequest, OAuth2AccessTokenResponse> getOnBehalfOfGrantCache() {
+        return onBehalfOfGrantCache;
+    }
+
+    public void setOnBehalfOfGrantCache(Cache<OnBehalfOfGrantRequest, OAuth2AccessTokenResponse> onBehalfOfGrantCache) {
+        this.onBehalfOfGrantCache = onBehalfOfGrantCache;
+    }
```

Comment:
Ikke at det nødvendigvis er feil altså :) Men bare lurer: Hvorfor settere på cachen i stedet for final-verdier som gis (som potensielt null) i constructor?  Vil det være aktuelt å bytte ut disse i fart?

Reply:
Mest for testbarhet (enklere å bytte ut i en test) og for å gjøre koden som bygger opp config for OAuth2AccessTokenService enklere mtp at det er "opt-in" for cache. Men har hatt det som ctor parametere tidligere, endte med settere til slutt, nå er jeg ikke så sikker igjen ;)

Reply:
hehe ;)  ikke heeelt sikker jeg heller.. men, for den config-oppbyggern så ville det vel blitt å først initialisere verdiene med null da, også eventuelt populere de ved opt-in, for _så_ å konstruere servicen.... og de byttes jo ikke ut i nåværende tester hvertfall ser jeg :)
Tenkte bare oppførselen kanskje er litt mer forutsigbar når de bare kan sette én gang, men å eventuelt endre senere blir jo ikke "breaking changes" for de som eventuelt tar i bruk spring-wrapperen heller da, så kan jo gruble litt fremover ;)